src/benchmarks/pmemobj_tx_add_range.cpp

```diff
@@ -0,0 +1,241 @@
+#include <cerrno>
+#include <cstddef>
+#include <cstdio>
+#include <cstdlib>
+#include <cstring>
+#include <fcntl.h>
+#include <sys/file.h>
+#include <sys/mman.h>
+#include <unistd.h>
+
+#include "benchmark.hpp"
+#include "file.h"
+#include "libpmemobj.h"
+
+#define LAYOUT_NAME "tx_add_range_benchmark"
+
+/*
+ * RANGE_COEFFICIENT -- pool has to hold every allocated object with
+ * its snapshot (1 + 1), plus 0.5 because of fragmentation
+ */
+#define RANGE_COEFFICIENT (1 + 1 + 0.5)
+
+/*
+ * obj_bench_args -- benchmark specific command line options
+ */
+struct obj_bench_args {
+	uint64_t nranged; /* number of allocated objects */
+};
+
+/*
+ * obj_bench -- benchmark context
+ */
+struct obj_bench {
+	PMEMobjpool *pop; /* persistent pool handle */
+	PMEMoid *ranged;  /* array of allocated objects */
+	size_t obj_size;  /* size of allocated object */
+	uint64_t nranged; /* number of allocated objects */
+};
+
+/*
+ * init_objects -- allocate persistent objects
+ */
+static int
+init_objects(struct obj_bench *ob)
+{
+	assert(ob->nranged != 0);
+	ob->ranged = (PMEMoid *)malloc(ob->nranged * sizeof(PMEMoid));
+
+	if (!ob->ranged) {
+		perror("malloc");
+		return -1;
+	}
+
+	for (uint64_t i = 0; i < ob->nranged; i++) {
+		PMEMoid *oid = &ob->ranged[i];
+		if (pmemobj_alloc(ob->pop, oid, ob->obj_size, 0, nullptr,
+				  nullptr)) {
+			perror("pmemobj_alloc");
+			return -1;
+		}
+	}
+
+	return 0;
+}
+
+/*
+ * tx_add_range_init -- initialization function
+ */
+static int
+tx_add_range_init(struct benchmark *bench, struct benchmark_args *args)
+{
+	assert(bench != nullptr);
+	assert(args != nullptr);
+	assert(args->opts != nullptr);
+
+	struct obj_bench_args *bargs = (struct obj_bench_args *)args->opts;
+
+	enum file_type type = util_file_get_type(args->fname);
+	if (type == OTHER_ERROR) {
+		fprintf(stderr, "could not check type of file %s\n",
+			args->fname);
+		return -1;
+	}
+
+	auto *ob = (struct obj_bench *)malloc(sizeof(struct obj_bench));
+	if (ob == nullptr) {
+		perror("malloc");
+		return -1;
+	}
+	pmembench_set_priv(bench, ob);
+
+	size_t psize;
+
+	if (args->is_poolset || type == TYPE_DEVDAX)
+		psize = 0;
+	else {
+		psize = bargs->nranged * args->dsize * RANGE_COEFFICIENT;
+		if (psize < PMEMOBJ_MIN_POOL)
+			psize = PMEMOBJ_MIN_POOL;
+	}
+
+	/* create pmemobj pool */
+	ob->pop = pmemobj_create(args->fname, LAYOUT_NAME, psize, args->fmode);
+	if (ob->pop == nullptr) {
+		fprintf(stderr, "%s\n", pmemobj_errormsg());
+		goto err;
+	}
+
+	ob->nranged = bargs->nranged;
+	ob->obj_size = args->dsize;
+
+	if (init_objects(ob))
+		goto err_pop_close;
+
+	return 0;
+
+err_pop_close:
+	pmemobj_close(ob->pop);
+err:
+	free(ob);
+	return -1;
+}
+
+/*
+ * tx_add_range_op -- actual benchmark operation
+ */
+static int
+tx_add_range_op(struct benchmark *bench, struct operation_info *info)
+{
+	auto *ob = (struct obj_bench *)pmembench_get_priv(bench);
+	int ret = 0;
+
+	TX_BEGIN(ob->pop)
+	{
+		for (uint64_t i = 0; i < ob->nranged; i++) {
```

Comment:
it would be useful to have non-sequential pattern (one way to implement this would be to shuffle the ranged array).